src/analyz/ty_json.rs

```diff
@@ -662,329 +658,108 @@ impl<'tcx> ToJson<'tcx> for ty::subst::GenericArg<'tcx> {
     }
 }
 
-fn do_const_eval<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    def_id: DefId,
-    substs: ty::subst::SubstsRef<'tcx>,
-    promoted: Option<mir::Promoted>,
-) -> mir::interpret::ConstValue<'tcx> {
-    let param_env = ty::ParamEnv::reveal_all();
-    let instance = ty::Instance::resolve(tcx, param_env, def_id, substs).unwrap();
-    let cid = interpret::GlobalId {
-        instance,
-        promoted,
-    };
-    tcx.const_eval_validated(param_env.and(cid)).unwrap()
-}
-
 fn eval_array_len<'tcx>(
     tcx: TyCtxt<'tcx>,
-    c: &'tcx ty::Const<'tcx>,
+    c: ty::Const<'tcx>,
 ) -> usize {
-    let evaluated = match c.val {
-        ty::ConstKind::Unevaluated(def_id, substs, promoted) => {
-            do_const_eval(tcx, def_id, substs, promoted)
-        },
-        ty::ConstKind::Value(val) => val,
-        _ => panic!("don't know how to translate ConstKind::{:?}", c.val),
-    };
-    match evaluated {
-        interpret::ConstValue::Scalar(interpret::Scalar::Raw { size: _, data }) => {
-            assert!(data <= usize::MAX as u128);
-            data as usize
-        },
-        _ => panic!("impossible: array size is not a scalar?"),
+    match c.kind() {
+        // ty::ConstKind::Unevaluated(un) => {
+        //     tcx.const_eval_resolve(ty::ParamEnv::reveal_all(), un.expand(), None).unwrap()
+        // },
+        // NB: the type of "val" changed to valtree - try to get the size out of that
+        ty::ConstKind::Value(ty::ValTree::Leaf(val)) =>
+            val.try_to_machine_usize(tcx).expect("expecting usize value from constant") as usize,
+        ref val => panic!("don't know how to translate ConstKind::{:?}", val),
```

Comment:
Could you add some tests in crux-mir that try to exercise the `Unevaluated` case?  It looks like we don't have any at the moment.

```Rust
const FOO: usize = 5;

fn f(xs: [i32; FOO]) -> i32 {
    xs[2]
}


trait HasSize {
    const SIZE: usize;
}

impl HasSize for () {
    const SIZE: usize = 6;
}

fn g<T: HasSize>(xs: [i32; T::SIZE]) -> i32 {
    xs[2]
}

// Do something with `f` and `g::<()>`...
```

If these pass, then the commented-out case can be safely removed.  (And if they don't, we can mark them as expected-fail / known bug and remove the case anyway.)

Reply:
Does this code actually work? When I try it, I get:

```
$ rustc test/conc_eval/consts/unevaluated.rs
error[E0599]: no associated item named `SIZE` found for type parameter `T` in the current scope
  --> test/conc_eval/consts/unevaluated.rs:16:31
   |
16 | fn g<T: HasSize>(xs: [i32; T::SIZE]) -> i32 {
   |                               ^^^^ associated item not found in `T`
   |
   = help: items from traits can only be used if the type parameter is bounded by the trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
```

(I might be missing something.)

Reply:
Ah, yep, the second one is not (yet) allowed by rustc.  I guess it should look about the same as the first one after mir-json is done monomorphizing, so just testing with `f` should be sufficient.

Reply:
I've added an `unevaluated.rs` test case in https://github.com/GaloisInc/crucible/pull/1096 that tests `f` above.